Sources/Views/MessageLabel.swift

```diff
@@ -22,18 +22,552 @@
+            if !isConfiguring { setNeedsDisplay() }
+        }
+    }
+    
+    public var lineBreakMode: NSLineBreakMode = .byWordWrapping {
+        didSet {
+            textContainer.lineBreakMode = lineBreakMode
+            if !isConfiguring { setNeedsDisplay() }
+        }
+    }
+    
+    public var lineFragmentPadding: CGFloat = 0 {
+        didSet {
+            textContainer.lineFragmentPadding = lineFragmentPadding
+            if !isConfiguring { setNeedsDisplay() }
+        }
+    }
+    
+    public var numberOfLines: Int = 0 {
+        didSet {
+            textContainer.maximumNumberOfLines = numberOfLines
+            if !isConfiguring { setNeedsDisplay() }
+        }
+    }
+    public var textAlignment: NSTextAlignment = .center {
+        didSet {
+            textStorage.addAttribute(key: .paragraphStyle, value: paragraphStyle)
+            if !isConfiguring { setNeedsDisplay() }
+        }
+    }
 
+    public var textInsets: UIEdgeInsets = .zero {
+        didSet {
+            if !isConfiguring { setNeedsDisplay() }
+        }
+    }
+    
+    public var enabledDetectors: [DetectorType] = [] {
+        didSet {
+            setStorageByParsing(text: attributedText) {
+                if !self.isConfiguring { self.setNeedsDisplay() }
+            }
+        }
+    }
+    
+    public var phoneAttributes: [NSAttributedStringKey: Any] = [:] {
+        didSet {
+            updateDetectorAttributes(for: .phoneNumber)
+            if !isConfiguring { setNeedsDisplay() }
+        }
+    }
+    
+    public var addressAttributes: [NSAttributedStringKey: Any] = [:] {
+        didSet {
+            updateDetectorAttributes(for: .address)
+            if !isConfiguring { setNeedsDisplay() }
+        }
+    }
+    
+    public var urlAttributes: [NSAttributedStringKey: Any] = [:] {
+        didSet {
+            updateDetectorAttributes(for: .url)
+            if !isConfiguring { setNeedsDisplay() }
+        }
+    }
+    
+    public var dateAttributes: [NSAttributedStringKey: Any] = [:] {
+        didSet {
+            updateDetectorAttributes(for: .date)
+            if !isConfiguring { setNeedsDisplay() }
+        }
+    }
+    
+    // MARK: - Properties [Private]
+    
+    private var layoutManager: NSLayoutManager
+    
+    private var textContainer: NSTextContainer
+    
+    private var textStorage: NSTextStorage
+    
+    private var phoneResults: [NSRange: String?] = [:]
+    
+    private var addressResults: [NSRange: Any?] = [:]
+    
+    private var dateResults: [NSRange: Date?] = [:]
+    
+    private var urlResults: [NSRange: URL?] = [:]
+    
+    private var isConfiguring: Bool = false
+    
+    private var paragraphStyle: NSParagraphStyle {
+        
+        guard let text = attributedText else { return NSParagraphStyle() }
+        
+        var range = NSRange(location: 0, length: 0)
+        let existingStyle = text.attribute(.paragraphStyle, at: 0, effectiveRange: &range) as? NSMutableParagraphStyle
```

Comment:
Crashing -- docs explicitly mentions a crash. Need to figure this out. May have to apply it in a separate method